### PR TITLE
Fix #1518: Move root route configuration to routes.js, add documentation for creating a global saga

### LINKED ISSUE #1518: How to configure App level saga?
At the `rootRoute` configuration in `app.js`, App is passed in as the base component for `react-router`, and the `childRoutes` inside `routes.js` are the JavaScript representation of the `react-router` route configuration where sagas are injected. How can I inject a saga at the `rootRoute` level?

What I'm trying to accomplish is the ability to load information from a server when the user enters the app, no matter which route they enter. I'm trying to avoid adding a `componentWillMount` to every container.

I've modified my `containers/App/index.js` to connect to react-redux, and I've added a `global` reducer in in `reducers.js`

**containers/App/index.js**
```
componentWillMount() {
  this.props.onLoadAccountDetails();
}
...
export function mapDispatchToProps(dispatch) {
  return {
    onLoadAccountDetails: () => {
      dispatch(loadAccountDetails());
    },
    dispatch,
  };
}
export default connect(mapStateToProps, mapDispatchToProps)(App);
```

**reducers.js**
```
...
import globalReducer from 'upside-rewards-common/lib/containers/App/reducer';
...
export default function createReducer(asyncReducers) {
  return combineReducers({
    route: routeReducer,
    global: globalReducer,
    language: languageProviderReducer,
    ...asyncReducers,
  });
}
```

### PR DESCRIPTION
**Pull request by RobertSheaO**
Fix for issue #1518. The route configuration was previously split between `app.js` and `routes.js`. I've moved the `App` (rootRoute) configuration to the `routes.js` file, so it is all in one place. I also changed loading the root route to use `getComponent`, instead of `component`, to make it easier for people to implement a global saga if they want to do that. There is also new documentation on how to add a global saga. I've also updated the spacing for the route generators and tested them.

Original Pull Request Here: https://github.com/react-boilerplate/react-boilerplate/pull/1538

### CODE REVIEW

**Comment by gihrig** [comment]
@RobertSheaO Thanks for cleaning up the PR! ✨ 

I'll be watching the discussion in #1518.

**Comment by KarandikarMihir** [comment]
This is awesome. Quick question- why is there no saga in `App/`? I had to remove `sagas` and `injectSagas(sagas.default);` from the route config. Precisely, [from here](https://github.com/RobertSheaO/react-boilerplate/blob/bb69de2e18fd690560f613bef001b1d606f8df37/app/routes.js#L65) to get it to work properly. 

I love the overall idea. Although, there seems to be a [working solution](https://github.com/react-boilerplate/react-boilerplate/issues/1277#issuecomment-275651735) for this particular problem, this PR solves two problems:
1. Creates a good, if not ideal, example for developers of implementing app-level sagas. 
2. Unifies route configuration in a single file.

That being said, I haven't studied other implementations. There's a [nice comment](https://github.com/justingreenberg) from @justingreenberg but that involves installing `redux-saga-router`. 

I'd love to know what others think of this.

**Comment by RobertSheaO** [comment]
Ah, good catch. I'll update the PR with that change. As for the missing
app/sagas.js, I didn't want to put it in the base app, as not everyone will
need it.

**Comment by KarandikarMihir** [comment]
@RobertSheaO No problem. Great job!

**Comment by RobertSheaO** [comment]
@KarandikarMihir I agree that a built-in usage example would be great. Not sure about repo-loading, since that only happens on submit; however, I could perhaps add an example of preloading @mxstbr 's repos on app load. Thoughts?

**Comment by gihrig** [comment]
> There's a nice comment from @justingreenberg

@KarandikarMihir The "nice comment" link got changed to JK's profile page ???

>preloading @mxstbr 's repos on app load. Thoughts?

@RobertSheaO How about pulling the current GH star count for react-boilerplate?

**Comment by RobertSheaO** [comment]
Sounds good, I'll add that and update my PR

**Comment by marcandrews** [comment]
Nice work!

**Comment by KarandikarMihir** [comment]
@gihrig Sorry! Here's the [correct link](https://github.com/react-boilerplate/react-boilerplate/pull/1315#issuecomment-267888341)

**Comment by gihrig** [comment]
There are a number of open issues surrounding difficulties with sagas and the way they are handled.

Ref: #1384, #1420, #1315, #1518, #1447 and this PR #1545.

I'm concerned about the lack of real-world experience behind these various proposals. The only two I know of having been applied are #1420 used in production by @adelsz and and #1447 used in production by @jeffbski, but not with react-boilerplate. @kopax has experience with [#1447Comment](https://github.com/react-boilerplate/react-boilerplate/issues/1447#issuecomment-273010133) and react-boilerplate. Perhaps he can update us with his experience so far.

I'm personally leaning toward #1447 (redux-logic) and possibly some combination of other proposals, but have not gotten far enough into this to make firm stand. If anyone is interested, I have updated @jeffbski's [fork](https://github.com/jeffbski/react-boilerplate-logic/tree/dev) to the 3.4 react-boilerplate release.

What I would really like to see, for is basic requirements of a simple addition to the example app to be drawn up, such as implementation of a simple login facility for example, then implemented in each proposal. I think the experience of those implementing the same set of requirements in the several proposed changes to RBP async handling would be enlightening.

**Comment by RobertSheaO** [comment]
I've been busy these past few days; however, I'll whip up an example to put into the example app. I'll also read over the above linked issues to see how they impact what I'm doing. Personally, I like the simplicity of this solution, since it uses sagas in the same exact way as all the other sagas. For my particular use case, this is how I am using them:

1) Every route, except my login route, has an `onEnter` property that checks for a user token.
######routes.js
```JS
...
const { injectReducer, injectSagas, redirectToLogin } = getAsyncInjectors(store);
...
{
  path: '/',
  name: 'home',
  onEnter: redirectToLogin,
  getComponent(nextState, cb) {
  ...
  },
}
```

######asyncInjectors.js
```JS
function redirectToLogin(store) {
  return (nextState, replaceState) => {
    // Look into store for token, if invalid, replaceState with the `/login` path.
}
```

2) Inside my `App/index.js` file that wraps the entire app, I check the token validity again, and if it's valid and I don't already have the data, I send the request to retrieve data using the global saga.

######App/index.js
```JS
componentWillMount() {
  if (accessTokenIsValid) {
    if (!this.props.accountDetails) this.props.onLoadAccountDetails();
  }
}
```

**Comment by RobertSheaO** [comment]
@marcandrews See my above comment about using `onEnter` (react-router) for authentication. This happens at the earliest point in the app, the route config, which could potentially prevent executing unnecessary code for users who are not authenticated.

**Comment by marcandrews** [comment]
In the hopes of not detracting further away from @RobertSheaO and the awesome work he's done in this PR, I've created a separate issue #1554 in response to @gihrig's comment.

**Comment by jackywhlee** [comment]
I've implemented this PR on my application because I needed the ability to logout via sagas, and I did not want to import 1 saga onto every component onto my app, and this solution works nicely for me.

File: docs/js/global-saga.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+Next we'll import the `sagas.js` file into our `App` route configuration by updating our `getComponent` function to asynchronosly load our files. You can just replace it with the one below.
+######app/routes.js
+```JS
+export default function createRoutes(store) {
+  // create reusable async injectors using getAsyncInjectors factory
+  const { injectReducer, injectSagas } = getAsyncInjectors(store);
+
+  childRoutes : [
```
Comment (avdeev) [inline docs/js/global-saga.md]: `childRoutes = [`

File: docs/js/global-saga.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+## Usage
+
+We'll be doing the following:
+
+1. Adding `sagas.js` to `app/containers/App` directory
+2. importing `sagas.js` into the App container route configuration
+
+Like other sagas, your global saga will live within a container, the `App` container. Go ahead and add a file and use the following starter code; it's the same starter code for other sagas.
+
+######app/containers/App/sagas.js
```
Comment (avdeev) [inline docs/js/global-saga.md]: `###### app/containers/App/sagas.js`

File: docs/js/global-saga.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+];
+
+// Individual exports for testing
+export function* sagaName() {
+
+}
+```
+
+Next we'll import the `sagas.js` file into our `App` route configuration by updating our `getComponent` function to asynchronosly load our files. You can just replace it with the one below.
+######app/routes.js
```
Comment (avdeev) [inline docs/js/global-saga.md]: `###### app/routes.js`

**Comment by magiccookie** [comment]
please merge

**Comment by mxstbr** [comment]
Deploy preview failed.

Built with commit 2e577930bf285a99b2eb997fa884d665315eb9d1

https://app.netlify.com/sites/react-boilerplate/deploys/59690735a700c4681563d721

**Comment by justingreenberg** [comment]
@RobertSheaO thank you so much for your contribution :) as we have just upgraded to RR v4 and refactored `asyncInjectors`, this should be resolved (please see #1746) — i would love your feedback if you have the time to pull down `dev` branch and take it for a spin! thanks again

**Comment by RobertSheaO** [comment]
I’ll take a look. Thank you.

**Comment by circuitry2** [comment]
What happened to this? docs/js/global-saga..md no longer exists??? What is the recommended approach now?

**Pull request closed without merging**